zcash_client_sqlite/src/wallet.rs

```diff
@@ -41,63 +47,68 @@ pub mod transact;
 /// };
 ///
 /// let data_file = NamedTempFile::new().unwrap();
-/// let db = WalletDB::for_path(data_file).unwrap();
-/// let addr = get_address(&db, &Network::TestNetwork, AccountId(0));
+/// let db = WalletDB::for_path(data_file, Network::TestNetwork).unwrap();
+/// let addr = get_address(&db, AccountId(0));
 /// ```
 pub fn get_address<P: consensus::Parameters>(
-    data: &WalletDB,
-    params: &P,
+    wdb: &WalletDB<P>,
     account: AccountId,
 ) -> Result<Option<PaymentAddress>, SqliteClientError> {
-    let addr: String = data.0.query_row(
+    let addr: String = wdb.conn.query_row(
         "SELECT address FROM accounts
         WHERE account = ?",
         &[account.0],
         |row| row.get(0),
     )?;
 
-    decode_payment_address(params.hrp_sapling_payment_address(), &addr)
-        .map_err(|e| SqliteClientError(e.into()))
+    decode_payment_address(wdb.params.hrp_sapling_payment_address(), &addr)
+        .map_err(SqliteClientError::Bech32)
 }
 
 pub fn get_extended_full_viewing_keys<P: consensus::Parameters>(
-    data: &WalletDB,
-    params: &P,
-) -> Result<Vec<ExtendedFullViewingKey>, SqliteClientError> {
+    wdb: &WalletDB<P>,
+) -> Result<HashMap<AccountId, ExtendedFullViewingKey>, SqliteClientError> {
     // Fetch the ExtendedFullViewingKeys we are tracking
-    let mut stmt_fetch_accounts = data
-        .0
-        .prepare("SELECT extfvk FROM accounts ORDER BY account ASC")?;
+    let mut stmt_fetch_accounts = wdb
+        .conn
+        .prepare("SELECT account, extfvk FROM accounts ORDER BY account ASC")?;
 
     let rows = stmt_fetch_accounts
         .query_map(NO_PARAMS, |row| {
-            row.get(0).map(|extfvk: String| {
+            let acct = row.get(0).map(AccountId)?;
+            let extfvk = row.get(1).map(|extfvk: String| {
                 decode_extended_full_viewing_key(
-                    params.hrp_sapling_extended_full_viewing_key(),
+                    wdb.params.hrp_sapling_extended_full_viewing_key(),
                     &extfvk,
                 )
-                .map_err(|e| Error::Bech32(e))
-                .and_then(|k| k.ok_or(Error::IncorrectHRPExtFVK))
-                .map_err(SqliteClientError)
-            })
+                .map_err(SqliteClientError::Bech32)
+                .and_then(|k| k.ok_or(SqliteClientError::IncorrectHRPExtFVK))
+            })?;
+
+            Ok((acct, extfvk))
         })
         .map_err(SqliteClientError::from)?;
 
-    rows.collect::<Result<Result<_, _>, _>>()?
+    let mut res: HashMap<AccountId, ExtendedFullViewingKey> = HashMap::new();
+    for row in rows {
+        let (account_id, efvkr) = row?;
+        res.insert(account_id, efvkr?);
+    }
```

Comment:
I'm pretty sure you can `collect()` an iterator of `(key, value)` tuples into a `HashMap` 🙂

Reply:
It's all the error handling that I couldn't figure out here with `collect()`.

My kingdom for `traverse` (https://hackage.haskell.org/package/base-4.14.1.0/docs/Prelude.html#v:traverse)